README
OpenGL Building Blocks (oglbb)
by Scott Rubin

This program is my term project for Computer Graphics 2

This program allows the user to create 3d models consisting solely of cubes.  It
has many features including zoom, pan, 32 bit color, save and load and more. 

The gui is simple and self explanatory, I hope. It is made using gtk+-2.0, 
gtkglext and OpenGL. To run the program do the following

unzip and untar the source
since you are reading this you must have done this already

run "make"

run "./oglbb"

That's it! This program has been tested and found to work with the following

gtk+-2.0
gtkglext-1.0
Solaris 9
Linux 2.6.0
gcc 3.3
NVidia Cg Toolkit


Contents of this directory:

CHANGELOG - what was done in each version of the program
COPYING - The License
Makefile - The Makefile
RCS - The source repository
README - This file
*.c, *.h, *.cg - source files
oglbb - oglbb executable
shader - shader executable
samples/*.bmd - example models


CHANGELOG
next release (1.1):

.rib and .xml output both fully functioning
keyboard controls

current release (1.0):

First release! The following features are fully functional

saving and loading
modeling
rotate
pan
zoom
grid
origin
light toggle
saving and loading


shader.c
/* shader.c
 *
 * NVidia Cg Shader program for oglbb
 *
 * Open GL Building Blocks
 * Copyright 2004 Scott Rubin
 *
 * This file is part of oglbb.
 *
 * oglbb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * oglbb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with oglbb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
 
#include <stdio.h>
#include <stdlib.h>

#include <GL/glut.h>

#include <Cg/cg.h>
#include <Cg/cgGL.h>

#include "oglfunc.h"

#define MAX_BRICKS 1024


int numbricks;
struct brick model[MAX_BRICKS];
float zoom = 0.05;

GLfloat lightpos[] = { 3.0, 3.0, 3.0, 0.0 };
GLfloat lightN[] = { 0.0, 0.0, 0.0, 0.0 };
GLfloat lightY[] = { 1.0, 1.0, 1.0, 0.0 };

/* Cg Variables */
static CGcontext CgCtxt = NULL;
static CGprogram CgProg = NULL;
static CGprofile CgProf = CG_PROFILE_VP20;

static CGparameter ModelViewProjParam = NULL;
static CGparameter LightAngle = NULL;
static CGparameter LightY = NULL;

float degrees = 0.0;
float ypos = 0.0;

/* COPIES CONTENTS OF FILE INTO MODEL */
void readfile( const char *filename ){
	
	struct brick retval[MAX_BRICKS];
	
	FILE *fp = fopen( filename, "r");
	
	int n=0;
	
	if( fp == NULL ){
		fprintf(stderr, "Cant open file %s", filename );
		exit(0);
	}else{
		while( fscanf( fp, "%d %d %d %d %d %d",
		&retval[n].x, &retval[n].y, &retval[n].z, 
		&retval[n].r, &retval[n].g, &retval[n].b ) != EOF ){
			n++;
		}
	}
	
	numbricks = n;
	for( int p=0; p < n; p++ ){
		model[p].x = retval[p].x;
		model[p].y = retval[p].y;
		model[p].z = retval[p].z;
		model[p].r = retval[p].r;
		model[p].g = retval[p].g;
		model[p].b = retval[p].b;
	}
}

void setCgParams(){
	
	if( ModelViewProjParam != NULL )
		cgGLSetStateMatrixParameter( ModelViewProjParam,
								CG_GL_MODELVIEW_PROJECTION_MATRIX,
								CG_GL_MATRIX_IDENTITY );
	
	if( LightAngle != NULL )
		cgGLSetParameter1f( LightAngle, degrees );
		
	if( LightY != NULL )
		cgGLSetParameter1f( LightY, ypos );
	
}

/* draw all the cubes */
void drawStuff(){

	glRotatef( 5 * degrees, 0.0, 1.0, 0.0 );
	cgGLBindProgram( CgProg );
	setCgParams();
	cgGLEnableProfile( CgProf );	
	
	glPushMatrix();

	for( int n=0; n < numbricks; n++ ){
	
		float r = (float)model[n].r;
		float g = (float)model[n].g;
		float b = (float)model[n].b;
		r = r / 255.0;
		g = g / 255.0;
		b = b / 255.0;
		float x = model[n].x * zoom;
		float y = model[n].y * zoom;
		float z = model[n].z * zoom;
		
		glColor3f( r, g, b );
		
		glBegin( GL_QUADS );
			glNormal3f( 0.0, -1.0, 0.0 );				
			glVertex3f( x, y, z );
			glVertex3f( x+zoom, y, z );
			glVertex3f( x+zoom, y, z+zoom );
			glVertex3f( x, y, z+zoom );
		glEnd();
		glBegin( GL_QUADS );
			glNormal3f( -1.0, 0.0, 0.0 );
			glVertex3f( x, y, z );
			glVertex3f( x, y, z+zoom );
			glVertex3f( x, y+zoom, z+zoom );
			glVertex3f( x, y+zoom, z );
		glEnd();
		glBegin( GL_QUADS );
			glNormal3f( 0.0, 0.0, -1.0 );
			glVertex3f( x, y, z );
			glVertex3f( x+zoom, y, z );
			glVertex3f( x+zoom, y+zoom, z );
			glVertex3f( x, y+zoom, z );
		glEnd();
		glBegin( GL_QUADS );
			glNormal3f( 1.0, 0.0, 0.0 );
			glVertex3f( x+zoom, y, z );
			glVertex3f( x+zoom, y+zoom, z );
			glVertex3f( x+zoom, y+zoom, z+zoom );
			glVertex3f( x+zoom, y, z+zoom );
		glEnd();
		glBegin( GL_QUADS );
			glNormal3f( 0.0, 0.0, 1.0 );
			glVertex3f( x, y, z+zoom );
			glVertex3f( x, y+zoom, z+zoom );
			glVertex3f( x+zoom, y+zoom, z+zoom );
			glVertex3f( x+zoom, y, z+zoom );
		glEnd();
		glBegin( GL_QUADS );
			glNormal3f( 0.0, 1.0, 0.0 );
			glVertex3f( x, y+zoom, z );
			glVertex3f( x+zoom, y+zoom, z );
			glVertex3f( x+zoom, y+zoom, z+zoom );
			glVertex3f( x, y+zoom, z+zoom );
		glEnd();
	
	}
	glPopMatrix();
	
	cgGLDisableProfile( CgProf );
	

}

//The almighty display function
void display( void ) {
	
  	/* THE BOX */
	glViewport( 0, 0, 800, 600 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	gluOrtho2D( 0, 800, 0, 600 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	
	/* ANGULAR VIEWPORT */
	glViewport( 0, 0, 800, 600 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	glOrtho( -1.0, 1.0, -1.0, 1.0, 1, 256 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	gluLookAt( 5.0, 5.0, 5.0, 0.0, 0.0, 0.0, -1.0, 1.0, -1.0 );
	drawStuff();
	
	glutSwapBuffers();

}


void idle( void ){
	degrees += 0.00075;
	ypos += 0.0005;
	glutPostRedisplay();
}


/* Initialize glut */
void init( int argc, char *argv[] ) {

  glutInit( &argc, argv );
  glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
  glutInitWindowSize( 800, 600 );
  glutInitWindowPosition( 75, 75 );
  glutCreateWindow( "OpenGL Building Block Shader" );
  
  glutDisplayFunc( display );
  glutIdleFunc( idle );

  glClearColor( 1.0, 1.0, 1.0, 0.0 );
  glEnable( GL_DEPTH_TEST );
  
}

void CgInit(){
	CgCtxt = cgCreateContext();
	CgProg = cgCreateProgramFromFile( CgCtxt, CG_SOURCE, "shader.cg",
											CgProf, NULL, NULL );
	if( CgProg != NULL ){
		
		cgGLLoadProgram( CgProg );
		
		ModelViewProjParam = cgGetNamedParameter( CgProg, "ModelViewProj" );
		
		LightAngle = cgGetNamedParameter( CgProg, "LA" );
		
		LightY = cgGetNamedParameter( CgProg, "LY" );
	}
}

/* THE MAIN */
int main( int argc, char** argv ) {
	
	if( argc == 2 ){
		char *a = argv[1];
		readfile( a );
	}else{
		printf( "Usage: shader <filename>\n" );
		exit(0);
	}
	
	init(argc, argv);
	CgInit();
	glutMainLoop();
	
	cgDestroyProgram( CgProg );
	cgDestroyContext( CgCtxt );

	return 0;
}


oglfunc.c
/* oglfunc.c
 *
 * OpenGL functions
 *
 * Open GL Building Blocks
 * Copyright 2004 Scott Rubin
 *
 * This file is part of oglbb.
 *
 * oglbb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * oglbb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with oglbb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
 
#include <stdio.h>
#include <GL/gl.h>
#include <GL/glu.h>
#include <glib.h>
#include <gdk/gdk.h>
#include <gtk/gtk.h>
#include <gtk/gtkgl.h>
#include "oglfunc.h"

#define MAX_BRICKS 1024

/* OPENGL DATA */
float zoom;
struct brick model[MAX_BRICKS];
int numbricks;
struct brick current;
float rotatex, rotatey, rotatez;
float panx, pany, panz;
int grid;
int origin;

/* LIGHTS DEFINED */
int light = 1;
GLfloat lightpos[] = { 3.0, 3.0, 3.0, 0.0 };
GLfloat lightN[] = { 0.0, 0.0, 0.0, 0.0 };
GLfloat lightY[] = { 1.0, 1.0, 1.0, 0.0 };

/* CHEAP HACK TO ERASE ALL BRICKS */
/* However, it does work and does perform well */
void eraseall(){
	numbricks = 0;
}	

/* Put current at origin */
void movereset(){
	current.x = 0;
	current.y = 0;
	current.z = 0;
}

/* Undo all rotattions */
void rotatereset(){
	rotatex = 0;
	rotatey = 0;
	rotatez = 0;
}

/* Undo all panning */
void panreset(){
	panx = 0;
	pany = 0;
	panz = 0;
}

/* Set zoom to default */
void zoomreset(){
	zoom = 0.05;
}

/* RESET EVERYTHING! */
void resetall(){
	movereset();
	rotatereset();
	panreset();
	zoomreset();
}

/* moves the selector brick */
void moveBlock( int x, int y, int z ){
	current.x += x;
	current.y += y;
	current.z += z;
}

/* delete selected brick */
void clearBlock() {

	int brick = -1;
	for( int n=0; n < numbricks; n++ ){
		if( model[n].x == current.x &&
			model[n].y == current.y &&
			model[n].z == current.z ){
					brick = n;
		}
	}
	if( brick != -1 ){
		for( brick; brick < numbricks-1; brick++ ){
			model[brick] = model[brick+1];
		}
		numbricks--;
	}
}


/* places brick of appropriate color at selected space */
void setBlock(){
	int replace=0;
	if( numbricks < MAX_BRICKS ){
		for( int n=0; n < numbricks; n++ ){
			if( model[n].x == current.x &&
				model[n].y == current.y &&
				model[n].z == current.z ){
					model[n] = current;
					replace = 1;
			}
		}
		if( replace == 0 ){
			model[numbricks] = current;
			numbricks++;
		}
	}else{
		printf( "OUT OF BRICKS\n" );
	}
}

/* Draws the lights */
void drawLight(){
	glLightfv( GL_LIGHT0, GL_AMBIENT, lightY );
	glLightfv( GL_LIGHT0, GL_DIFFUSE, lightN );
	glLightfv( GL_LIGHT0, GL_SPECULAR, lightN );
	glLightfv( GL_LIGHT0, GL_POSITION, lightpos );
	
	glLightfv( GL_LIGHT1, GL_AMBIENT, lightN);
	glLightfv( GL_LIGHT1, GL_DIFFUSE, lightY );
	glLightfv( GL_LIGHT1, GL_SPECULAR, lightN );
	glLightfv( GL_LIGHT1, GL_POSITION, lightpos );
	
	glLightfv( GL_LIGHT2, GL_AMBIENT, lightN );
	glLightfv( GL_LIGHT2, GL_DIFFUSE, lightN );
	glLightfv( GL_LIGHT2, GL_SPECULAR, lightY );
	glLightfv( GL_LIGHT2, GL_POSITION, lightpos );
}

/* Draws a single cube */
void drawCube( struct brick b ){
	glPushMatrix();
		glColor3ub( b.r, b.g, b.b );
		glTranslatef( zoom * b.x, zoom *  b.y, zoom * b.z );
		gdk_gl_draw_cube( TRUE, zoom );
	glPopMatrix();
}

/* DRAW THE GRID */
void drawGrid(){
	glPushMatrix();
	glColor3ub( 255, 255, 255 );
	glBegin( GL_LINES );
		for( float x = 0; x < 1; x += zoom ){
			glVertex3f( x, 0, 0 );
			glVertex3f( x, 0, 1 );
		}
		for( float z = 0; z < 1; z += zoom ){
			glVertex3f( 0, 0, z );
			glVertex3f( 1, 0, z );
		}
	glEnd();
	glPopMatrix();
				
}

/* DRAW THE ORIGIN AXES */
void drawOrigin(){
	glPushMatrix();
	glColor3ub( 255, 0, 0 );
		glBegin( GL_LINES );
			glVertex3i( 0, 0, 0 );
			glVertex3i( 1, 0, 0 );
		glEnd();
		glColor3ub( 0, 255, 0 );
		glBegin( GL_LINES );
			glVertex3i( 0, 0, 0 );
			glVertex3i( 0, 1, 0 );
		glEnd();
		glColor3ub( 0, 0, 255 );
		glBegin( GL_LINES );
			glVertex3i( 0, 0, 0 );
			glVertex3i( 0, 0, 1 );
		glEnd();
	glPopMatrix();
}

/* draw all the cubes in the model, then the selector brick */
void drawStuff(){
	glTranslatef( panx, pany, panz );
	glPushMatrix();
	
	/* ROTATE */
	glRotatef( rotatex, 1.0, 0.0, 0.0 );
	glRotatef( rotatey, 0.0, 1.0, 0.0 );
	glRotatef( rotatez, 0.0, 0.0, 1.0 );
	
	/*Draw the selector block, always ambiently lighted */
	glDisable( GL_LIGHTING );
	if( origin == 0 ){
		drawOrigin();
	}
	if( grid == 0 ){
		drawGrid();
	}
	glPushMatrix();
		glColor3ub( 255, 255, 255 );
		glTranslatef( zoom * current.x, zoom * current.y, zoom * current.z );
		gdk_gl_draw_cube( FALSE, zoom );
		glColor3ub( current.r, current.g, current.b );
		gdk_gl_draw_cube( TRUE, zoom );
	glPopMatrix();
	glEnable( GL_LIGHTING );
	
	/* Draw all the other blocks in the model */
	glPushMatrix();
		for( int x=0; x < numbricks; x++ ){
			drawCube( model[x] );
		}
	glPopMatrix();
	glPopMatrix();
}

/* ALMIGHTY DISPLAY FUNCTION!!! */
void display(){

	/* TURN OFF THAT LIGHT! */
	glDisable( GL_LIGHTING );
	
	glViewport( 0, 0, 800, 600 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	gluOrtho2D( 0, 800, 0, 600 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );
	glColor3ub( 255, 255, 255 );
	
	/* seperate the viewports with lines */
	glColor3ub( 255, 255, 255 );
	glBegin( GL_LINES );
		glVertex2i( 600, 0 );
		glVertex2i( 600, 600 );
	glEnd();
	glBegin( GL_LINES );
		glVertex2i( 600, 200 );
		glVertex2i( 800, 200 );
	glEnd();
	glBegin( GL_LINES );
		glVertex2i( 600, 400 );
		glVertex2i( 800, 400 );
	glEnd();
	
	/* LET THERE BE LIGHT! */
	glEnable( GL_LIGHTING );
	
	if( light == 0 ){
		glEnable( GL_LIGHT0 );
		glDisable( GL_LIGHT1 );
		glDisable( GL_LIGHT2 );
	}else{
		glDisable( GL_LIGHT0 );
		glEnable( GL_LIGHT1 );
		glEnable( GL_LIGHT2 );
	}
	drawLight();
	
	/* FRONT VIEWPORT */
	glViewport( 600, 400, 200, 200 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	glOrtho( -1.0, 1.0, -1.0, 1.0, 1, 256 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	gluLookAt( 0.5, 0.5, 3.0, 0.5, 0.5, 0.5, 0.0, 1.0, 0.0 );
	drawStuff();

	/* TOP VIEWPORT */
	glViewport( 600, 200, 200, 200 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	glOrtho( -1.0, 1.0, -1.0, 1.0, 1, 256 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	gluLookAt( 0.5, 3.0, 0.5, 0.5, 0.5, 0.5, 0.0, 0.0, -1.0 );
	drawStuff();
	
	/* SIDE VIEWPORT */
	glViewport( 600, 0, 200, 200 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	glOrtho( -1.0, 1.0, -1.0, 1.0, 1, 256 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	gluLookAt( 3.0, 0.5, 0.5, 0.5, 0.5, 0.5, 0.0, 1.0, 0.0 );
	drawStuff();

	/* ANGULAR VIEWPORT (the big one) */
	glViewport( 0, 0, 600, 600 );
	glMatrixMode( GL_PROJECTION );
	glLoadIdentity();
	glOrtho( -1.0, 1.0, -1.0, 1.0, 1, 256 );
	glMatrixMode( GL_MODELVIEW );
	glLoadIdentity();
	gluLookAt( 5.0, 5.0, 5.0, 0.0, 0.0, 0.0, -1.0, 1.0, -1.0 );
	drawStuff();

}


oglfunc.h
/* oglfunc.h
 *
 * The brick structure
 *
 * Open GL Building Blocks
 * Copyright 2004 Scott Rubin
 *
 * This file is part of oglbb.
 *
 * oglbb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * oglbb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with oglbb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
 
#ifndef OGL_BRICK
#define OGL_BRICK
 
struct brick{
	int x, y, z, r, g, b;
};
 
#endif


blocker.c
/* blocker.c
 *
 * The MAIN file.
 * Contains gtk+ and gtkglext code.
 *
 * Open GL Building Blocks
 * Copyright 2004 Scott Rubin
 * 
 * Portions of this code were borrowed from the following sources:
 * examples in gtkglext cvs repository
 * gtk-2.0 tutorial
 *
 * This file is part of oglbb.
 *
 * oglbb is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * oglbb is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with oglbb; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
 
#ifdef G_OS_WIN32
#define WIN32_LENA_AND_MEAN 1
#include <windows.h>
#endif
 
#include "callbacks.c"
#include "icallbacks.c"

/* PULLDOWN MENU STRUCTURE*/
GtkItemFactoryEntry menu_items[] = {
  { "/_File",         NULL,         NULL,           0, "<Branch>" },
  { "/File/_New",     "<control>N", file_new,    0, "<StockItem>", GTK_STOCK_NEW },
  { "/File/_Open",    "<control>O", open_file,    0, "<StockItem>", GTK_STOCK_OPEN },
  { "/File/_Save",    "<control>S", save_file,    0, "<StockItem>", GTK_STOCK_SAVE },
  { "/File/_Quit",    "<CTRL>Q",    gtk_main_quit, 0, "<StockItem>", GTK_STOCK_QUIT },
};

/* NUMBER OF MENU ITEMS */
gint nmenu_items = sizeof (menu_items) / sizeof (menu_items[0]);

/* MENU WIDGET MAKER */
GtkWidget *get_menubar_menu( GtkWidget  *window ){

  GtkItemFactory *item_factory;
  GtkAccelGroup *accel_group;
  accel_group = gtk_accel_group_new();
  item_factory = gtk_item_factory_new( GTK_TYPE_MENU_BAR, "<main>", 
  										accel_group );
  gtk_item_factory_create_items( item_factory, nmenu_items, menu_items, NULL );
  gtk_window_add_accel_group( GTK_WINDOW (window), accel_group );
  return gtk_item_factory_get_widget( item_factory, "<main>" );
}

/* BUTTON MAKER FUNCTION */
void drawbutton( char *tx, char *sg, GtkWidget *btn, GtkWidget *tb, 
					GtkWidget *dar, int a, int b, int c, int d ){
	
	btn = gtk_button_new_with_mnemonic( tx );
	g_signal_connect( G_OBJECT( btn ), "clicked",
		      G_CALLBACK( callback ), (gpointer)sg );
	g_signal_connect_after( G_OBJECT( btn ), "clicked", 
				G_CALLBACK(refresh), (gpointer)dar );
	gtk_table_attach_defaults( GTK_TABLE( tb ), btn, a, b, c, d );
}

/* THE MAIN */
int main( int argc, char *argv[] ){

	/* GTK "OBJECTS" */
    GtkWidget *window;	//the main window
    GtkWidget *button;	//buttons
	GtkWidget *toggle;	//toggle buttons
    GtkWidget *big_vbox;//seperate the menu from other stuffs
	GtkWidget *hbox;	//seperate tool buttons and drawing area
	GtkWidget *vbox; 	//seperate button area
	GtkWidget *table;	//button area
	GtkWidget *label;	//labels
	GtkWidget *darea;   //drawing area
	GtkWidget *menubar; //pulldown menus
	
	/* OPENGL "OBJECTS" */
	GdkGLConfig *glconfig;
	gint major, minor;
	
	/* GTK INITIALIZATION */
    gtk_init( &argc, &argv );
	gtk_gl_init( &argc, &argv );

	/* OPENGL CONFIGURATION */
	gdk_gl_query_version( &major, &minor );
	glconfig = gdk_gl_config_new_by_mode( GDK_GL_MODE_RGB |
										  GDK_GL_MODE_DEPTH |
										  GDK_GL_MODE_DOUBLE );	
	if( glconfig == NULL ){
		glconfig = gdk_gl_config_new_by_mode( GDK_GL_MODE_RGB |
											  GDK_GL_MODE_DEPTH );
		if( glconfig == NULL ){
			g_print( "ERROR, NO OPENGL\n" );
		}
	}
	
	/* THE MAIN WINDOW */
    window = gtk_window_new( GTK_WINDOW_TOPLEVEL );
    gtk_window_set_title( GTK_WINDOW( window ), "OpenGL Building Blocks" );
	/*auto redraw window (for opengl)*/
	gtk_container_set_reallocate_redraws( GTK_CONTAINER( window ), TRUE);
    g_signal_connect( G_OBJECT( window ), "delete_event",
		      G_CALLBACK( delete_event ), NULL);
    gtk_container_set_border_width( GTK_CONTAINER( window ), 10 );
	gtk_window_set_resizable( GTK_WINDOW( window), FALSE );
	
	/* PULLDOWN MENU */
    big_vbox  = gtk_vbox_new( FALSE, 1 );
	gtk_container_set_border_width( GTK_CONTAINER( big_vbox ), 1 );
	gtk_container_add( GTK_CONTAINER (window), big_vbox );
	menubar = get_menubar_menu( window );
	gtk_box_pack_start( GTK_BOX( big_vbox ), menubar, FALSE, TRUE, 0 );
	
	/* PACKING AREAS */
	hbox = gtk_hbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX( big_vbox ), hbox, FALSE, FALSE, 0 );
	vbox = gtk_vbox_new( FALSE, 0 );
	gtk_box_pack_start( GTK_BOX( hbox ), vbox, FALSE, FALSE, 10 );
	
	/* OPENGL DRAWING AREA */
	darea = gtk_drawing_area_new();
	gtk_widget_set_size_request( darea, 800, 600 );
	gtk_widget_set_gl_capability( darea, glconfig, NULL, TRUE,
								  GDK_GL_RGBA_TYPE);
	g_signal_connect_after( G_OBJECT(darea), "realize",
							G_CALLBACK(realize), NULL );
	g_signal_connect( G_OBJECT(darea), "configure_event",
					  G_CALLBACK(configure_event), NULL );
	g_signal_connect( G_OBJECT(darea), "expose_event",
					  G_CALLBACK(expose_event), NULL );
		
	/* BUTTON TABLE */
	table = gtk_table_new( 17, 3, FALSE );
	gtk_box_pack_start( GTK_BOX( vbox ), table, FALSE, FALSE, 0 );
	
	/* LABELS */
	label = gtk_label_new( "Move" );
	gtk_misc_set_alignment( GTK_MISC( label ), 0, 1 );
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 2, 0, 1);
	gtk_widget_show( label );
	
	label = gtk_label_new( "Rotate" );
	gtk_misc_set_alignment( GTK_MISC( label ), 0, 1 );
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 2, 3, 4);
	gtk_widget_show( label );
	
	label = gtk_label_new( "Pan" );
	gtk_misc_set_alignment( GTK_MISC( label ), 0, 1 );
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 3, 6, 7);
	gtk_widget_show( label );
	
	label = gtk_label_new( "Zoom" );
	gtk_misc_set_alignment( GTK_MISC( label ), 0, 0 );
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 3, 9, 10);
	gtk_widget_show( label );
	
	label = gtk_label_new("");
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 3, 13, 14 );
	gtk_widget_show( label );
	
	label = gtk_label_new("Toggles");
	gtk_misc_set_alignment( GTK_MISC( label ), 0, 0 );
	gtk_table_attach_defaults( GTK_TABLE( table ), label, 0, 3, 11, 12);
	gtk_widget_show(label);
	
	/* MOVEMENT BUTTONS */
	drawbutton( "R", "mr", button, table, darea, 2, 3, 0, 1 );
	drawbutton( "X+", "mx+", button, table, darea, 0, 1, 1, 2 );
	drawbutton( "Y+", "my+", button, table, darea, 1, 2, 1, 2 );
	drawbutton( "Z+", "mz+", button, table, darea, 2, 3, 1, 2 );
	drawbutton( "X-", "mx-", button, table, darea, 0, 1, 2, 3 );
	drawbutton( "Y-", "my-", button, table, darea, 1, 2, 2, 3 );
	drawbutton( "Z-", "mz-", button, table, darea, 2, 3, 2 ,3 );
	
	/* ROTATION BUTTONS */
	drawbutton( "R", "rr", button, table, darea, 2, 3, 3, 4 );
	drawbutton( "X+", "rx+", button, table, darea, 0, 1, 4, 5 );
	drawbutton( "Y+", "ry+", button, table, darea, 1, 2, 4, 5 );
	drawbutton( "Z+", "rz+", button, table, darea, 2, 3, 4, 5 );
	drawbutton( "X-", "rx-", button, table, darea, 0, 1, 5, 6);
	drawbutton( "Y-", "ry-", button, table, darea, 1, 2, 5, 6 );
	drawbutton( "Z-", "rz-", button, table, darea, 2, 3, 5, 6 );
	
	/* PANNING BUTTONS */	
	drawbutton( "R", "pr", button, table, darea, 2, 3, 6, 7 );
	drawbutton( "X+", "px+", button, table, darea, 0, 1, 7, 8 );
	drawbutton( "Y+", "py+", button, table, darea, 1, 2, 7, 8 );
	drawbutton( "Z+", "pz+", button, table, darea, 2, 3, 7, 8 );
	drawbutton( "X-", "px-", button, table, darea, 0, 1, 8, 9 );
	drawbutton( "Y-", "py-", button, table, darea, 1, 2, 8, 9 );
	drawbutton( "Z-", "pz-", button, table, darea, 2, 3, 8, 9 );
		
	/* ZOOMING BUTTONS */
	drawbutton( "ZR", "zr", button, table, darea, 1, 2, 10, 11 );
	drawbutton( "Z+", "z+", button, table, darea, 0, 1, 10, 11 );
	drawbutton( "Z-", "z-", button, table, darea, 2, 3, 10, 11 );
	
	/* BIG BUTTONS */
	drawbutton( "Place _Block", "place", button, table, darea, 0, 3, 15, 16);
	drawbutton( "Delete Block", "delete", button, table, darea, 0, 3, 16, 17 );
	drawbutton( "Reset All", "reset", button, table, darea, 0, 3, 17, 18 );
	
	/* COLOR CHOOSING BUTTON */
	button = gtk_button_new_with_label( "Change Color" );
	g_signal_connect (G_OBJECT (button), "clicked",
		      G_CALLBACK (colorselect), (gpointer) "color");
	gtk_table_attach_defaults( GTK_TABLE (table), button, 0, 3, 14, 15 );
	gtk_widget_show( button );
	
	/* TOGGLE BUTTONS */	
	toggle = gtk_toggle_button_new_with_label("G");
	gtk_table_attach_defaults( GTK_TABLE( table ), toggle, 0, 1, 12, 13);
	gtk_widget_show( toggle );
	g_signal_connect (GTK_OBJECT (toggle), "toggled",
                     G_CALLBACK (toggle_button_callback), "G");
	g_signal_connect_after( G_OBJECT( toggle ), "toggled", 
				G_CALLBACK(refresh), (gpointer)darea );
	
	toggle = gtk_toggle_button_new_with_label("O");
	gtk_table_attach_defaults( GTK_TABLE( table ), toggle, 1, 2, 12, 13);
	gtk_widget_show( toggle );
	g_signal_connect (GTK_OBJECT (toggle), "toggled",
                     G_CALLBACK (toggle_button_callback), "O" );
	g_signal_connect_after( G_OBJECT( toggle ), "toggled", 
				G_CALLBACK(refresh), (gpointer)darea );
	
	toggle = gtk_toggle_button_new_with_label("L");
	gtk_table_attach_defaults( GTK_TABLE( table ), toggle, 2, 3, 12, 13);
	gtk_widget_show( toggle );
	g_signal_connect (GTK_OBJECT (toggle), "toggled",
                     G_CALLBACK (toggle_button_callback), "L");
	g_signal_connect_after( G_OBJECT( toggle ), "toggled", 
				G_CALLBACK(refresh), (gpointer)darea );
	
	gtk_box_pack_end( GTK_BOX( hbox ), darea, FALSE, FALSE, 0 ); 
	gtk_widget_show_all( window );
    gtk_main();

    return 0;
}
